### DISCUSSION TITLE
usage with FormProvider?

### DISCUSSION THREAD

**Discussion by bzbetty** [iway1/react-ts-form#31] [Q&A]
I'm used to using a <FormProvider /> so my inputs can useFormContext, but that needs a UseFormReturns<T> which doesn't seem to be supplied by this library.

On a sidenote this library seems to make a lot of use of Control which I thought was frowned upon directly accessing, could be because you seem to favour controlled and not uncontrolled form fields?

**Comment by iway1** [comment]
@bzbetty I use controls b/c that's how I've built my input components in the past and they're more flexible. (I'm a React Native developer mainly, so I have to use them.)  I'm not a huge fan of `register` in general b/c it doesn't cover all use cases. Rather just always use the API that I can build any input with.

With the controller approach, you get a lot of the same performance benefits without sacrificing flexibility. Sometimes you have to have a controlled component, and so going with `useController` means we can cover all use cases. Also, using the Control is the easiest way I know of to render custom error message components

I do think some (a lot?) of people prefer `register`. In theory should be pretty easy to add a `FormProvider` to the form component and allow using `useFormContext()`, though I'm not 100% sure I want to encourage the usage of `register` just because it's more to document and sort of expands the API of the library without adding functionality.

If you want, you can wrap your form in a `FormProvider`, and pass in the `UseFormReturn` to the `form` prop yourself, something like:

```tsx
const myForm = useForm();
return (
  <FormProvider {...myForm}>
    <Form 
      form={myForm}
    />
  </FormProvider>
)
```

Might add this, I'll have to think about it... Currently leaning towards just having one API `useTsController` for building inputs though.
